### DISCUSSION TITLE
Future steps

### DISCUSSION THREAD

**Discussion by Jefiozie** [bertrandg/angular-split#380] [Announcements]
# Greetings, Everyone! 🌟👋
 
 We are thrilled to unveil the roadmap for our project's next phase. This marks a wonderful opportunity for each member of our community to join the conversation, raise questions, and share insightful feedback about our future direction.
 
 ## 🛠️ Enhancing Maintainability
 Navigating the complexities of maintaining an open-source software (OSS) library is no small feat. To optimize our efforts, we're proposing pivotal changes to simplify contributions, streamline reviews, and improve our release processes.
 
 ### Streamlining CI Pipeline
 Our focus is on refining our Continuous Integration (CI) and Developer Experience (DX). With the [Optimizing CI Pipeline: Comprehensive Build, Lint, Test, and Automated Release Workflow #375](https://github.com/angular-split/angular-split/issues/375) initiative, we aim to bolster CI with additional checks. These enhancements are designed not only to facilitate smoother contributions but also to enrich the review process with more constructive feedback.
 
 ## 🌟 Project Evolution
 As Angular continues to evolve, it brings fresh opportunities. With the Angular CLI now supporting multi-project workspaces, we're eager to adapt our repository to align with these latest Angular CLI standards.
 
 ### Transitioning to Angular CLI Multi-Project Workspace
 Embracing the new Angular advancements, we're contemplating a shift to the Angular CLI Multi-Project Workspace. For more details, check out [Proposal for Migration to Angular CLI Multi-Project Workspace #379](https://github.com/angular-split/angular-split/issues/379).
 
 ## 👥 Fostering Community and Contributions
 Our community is the lifeblood of this project. In our commitment to reflect the values and contributions of our members, we're introducing new initiatives to recognize your efforts more effectively.
 
 ### Celebrating Contributions
 We're dedicated to acknowledging every contribution in a meaningful way. Through [Integrating 'All Contributors' Bot for Enhanced Contributor Recognition #371](https://github.com/angular-split/angular-split/issues/371), we aim to celebrate each and every contribution, regardless of its size.
 
 ### Upholding Our Code of Conduct
 In our journey, respect and inclusivity are non-negotiable. We're implementing a [code of conduct](https://github.com/angular-split/angular-split/issues/372) to ensure our community remains a respectful and safe environment for everyone. This code will serve as a guideline for all interactions and assist in resolving any issues that may arise.
 
 ---
 
 We are eager to delve into more initiatives, including standalone components, in the future. Your thoughts, questions, and feedback are invaluable as we embark on these exciting changes. Let's collaborate to elevate this project to new heights! 🚀💻

**Comment by Harpush** [comment]
Hey a few things:
1. I think the maintenance ideas are great and will indeed help DX.
2. I think using new angular features like standalone and signals is a great milestone to try and achieve.
3. I think we need to go over the open issues and see what we can do easily and what requires more thinking.
4. Lastly, I think we need to do some organizing to the existing code to make it more understandable and maintainable. That includes removing or changing features which are problematic or half developed if there are any.

**Reply by Jefiozie** [reply]
Thank you, @Harpush, for your valuable feedback.

We plan to address issues, considering closure, reopening, or checking for resolutions in newer versions. Any additional insights on this matter would be welcomed.

Regarding your last point, I agree that continuous improvement is essential. Personally, I find the "boy scout rule" a good approach for enhancing specific areas. Do you have examples or suggestions for code sections that might benefit from improvement?

**Reply by Harpush** [reply]
1. I think the split area should become a component and not a directive. Especially with the iframe fix integrated inside.
2. I think templates should have their own file.
3. I think the split component itself should be more ordered as it is a really large component with a lot of imperative code.
4. We should also look at features like collapse which I am yet to understand their value.

Just what I had in mind...

**Reply by SanderElias** [reply]
@Harpush Nice to meet again. 
The first order of business is improving the CI/CD story. To do that, we must first improve the DX. Please have a look at #370 for a first step.
We plan to do a "final" release in the current state but with all non-breaking PRs merged. (Unless there is a good reason that we are unaware of to _not_ merge any of those). 

The next step will be a breaking release, with the PR that is breaking merged but also completely moved to standalone.
If you have things that you think should be part of our first breaking release, please let us know; then, we can estimate the work and decide to include or postpone till the next breaking release.  I would rather bundle as much as possible in a single-breaking release than spread it out. At the same time, we are all very busy and we need to set realistic goals. Otherwise, nothing will ever get released.

**Comment by Harpush** [comment]
One more thing we should decide is pull request commits management. Whether we force push and order commits or just squash at the end so internal commits has no meaning outside the PR scope?

**Reply by Jefiozie** [reply]
In my view, it's crucial to condense commits so that those reaching the main branch are meaningful and align with the contribution guidelines. I'm considering implementing a Git hook to scrutinize commits, utilizing commitlint for validation. What do you think of this approach

**Reply by SanderElias** [reply]
I prefer to do squash&merge. I think it's silly to put an extra burden on maintainers for something that can be "fixed" while merging. 
The same goes for demanding specific formatting. Let's shoulder this "burden" as maintainers.
That being said, we should certainly make sure the merged PRs have proper formatting so automated tooling can take over from there. 
I'm not sure if there is a tool that can check the 'squash and merge` message for formatting. That would be the best solution in my eyes.

**Reply by Harpush** [reply]
I agree. I believe squash and merge is correct.
So generally the PR title should follow the commit guidelines and so the squash merge commit.
Does that mean "on the fly" commits don't have to follow specific guidelines? Or do we want to follow those guidelines in internal commits too but without merging commits and force pushing?

I am pretty sure commitlint can run only on specific branches but I am not sure if it will work or how to do it.

**Reply by SanderElias** [reply]
Jeffery found a tool to force the squash&merge msg format to follow conventions. That means we can allow anything, but we _need_ to fix the message before merging it. That will still allow automated changelogs and stuff, but without needing to demand strict rules on commit messages. 

I would say we should see if we can integrate this tool.

**Reply by SanderElias** [reply]
BTW, if we as maintainers use proper commit messages, the job during merge is a lot simpler. So it makes sense to still keep proper commit-message hygiene. 
I use the [Conventinal commits](https://marketplace.visualstudio.com/items?itemName=vivaxy.vscode-conventional-commits) plugin in vs-code, which makes it easier to format those things properly.

**Reply by Harpush** [reply]
So let's take a scenario - I am creating a feature branch and the first commit is:
`feat: add something`
Later I decide to add some stuff so:
`feat/refactor: add more something`
Later I have a bug so I commit:
`fix: fix PR comments`

At the end when you merge the branch you only take the first commit name as the squashed one:
`feat: add something`
Am I correct here? Or the fixup way is better?

**Reply by SanderElias** [reply]
@Harpush, the person who is going to press the `merge&commit` button, has to make sure the message is formatted correctly. Otherwise, the button will not work. (at least, that is what I understood from the tools docs.)
We prefer proper commit messages as that makes the job a bit easier, but we require nothing
